File: Java/0234_Palindrome_Linked_List.java
// 2020-08-04 23:09:28

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        int len = 0;
        for (ListNode p = head; p != null; p = p.next, ++len);

        ListNode cur = head, prev = null;
        for (int i = 0; i < len / 2; ++i) {
            ListNode next = cur.next;
            cur.next = prev;
            prev = cur;
            cur = next;
        }

        if ((len & 1) == 1) cur = cur.next;
        for (; prev != null && cur != null; prev = prev.next, cur = cur.next) {
            if (prev.val != cur.val) return false;
        }
        return true;
    }
}

File: C++/0234_palindromelinkedlist.cc
// 2020-08-05 00:07:24

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if (head == nullptr || head->next == null) return true;

        int len = 0;
        for (ListNode* p = head; p != nullptr; p = p->next) len += 1;

        ListNode* prev = nullptr;
        ListNode* cur = head;
        for (int i = 0; i < len / 2; i++) {
            ListNode* next = cur->next;
            cur->next = prev;
            prev = cur;
            cur = next;
        }
        if ((len & 1) == 1) cur = cur->next;
        for (; cur != nullptr && prev != nullptr; cur = cur->next, prev = prev->next) {
            if (cur->val != prev->val) return false;
        }

        return true;
    }
};
